### DISCUSSION TITLE
Best workflow for extension (hook) dev?

### DISCUSSION THREAD

**Discussion by TinyTurtleMedia** [directus/directus#10832] [Archive]
Hey all, brand new to Directus 9. I'm coming from headless Wordpress, so I know my way around PHP and bit of JS.

For my projects, I have to do quite a bit of calculations with fields inside a custom plugin. In Wordpress with PHP, this is somewhat similar to what I've seen in the Directus 9 documentation, except you build an extension. However, in Wordpress, you have a static php file in an active plugin directory, you enter some code, refresh the page, and the code either works or throws an error.

My issue is I can't seem to find any way to dev an extension without typing some code, building it, then moving it to the Directus 9 folder to see if it works or not, then doing that all over every time I need to make changes. The build seems to change the index.js syntax and all that. I assume there's a way to reference another file from within the index.js file that doesn't get changed when the extension is built?

I'm also not sure what the best practice is for writing data from viewing the documentation. I mostly understand how to retrieve data with the hooks and work with that in JS, but is there another way to then add it back into the database other than a POST request to the public API endpoints? I admit I haven't looked as hard into that part yet.

Any and all advice would be awesome, thanks!

**Comment by rijkvanzanten** [comment]
@nickrum Any up to date advice on the matter? 🙂

**Comment by nickrum** [comment]
> My issue is I can't seem to find any way to dev an extension without typing some code, building it, then moving it to the Directus 9 folder to see if it works or not, then doing that all over every time I need to make changes. The build seems to change the index.js syntax and all that. I assume there's a way to reference another file from within the index.js file that doesn't get changed when the extension is built?

When they have been built, App extensions always have to be single `index.js` files in order for the API to serve them to the browser. API extensions do not have to be built and can reference other files. They are easier to handle as single `index.js` files, though. That's why the Extensions SDK bundles them to a single file.
Unfortunately, the dev workflow is quite cumbersome right now. You could use the --watch flag to automatically rebuild each extension if the code changed and you could set the output folder to the extensions folder (either by setting directus:extension.path in package.json or by using the -o parameter). In addition to that, you could use nodemon to restart directus whenever one of the built extension entrypoints changes.
That being said, we are actively working on an "extension dev environment" to make all of this a lot easier. This is a tricky problem to solve though, so it will likely still take some time until it's done.

> I'm also not sure what the best practice is for writing data from viewing the documentation. I mostly understand how to retrieve data with the hooks and work with that in JS, but is there another way to then add it back into the database other than a POST request to the public API endpoints? I admit I haven't looked as hard into that part yet.

You can use the `ItemsService` to do this:

```js
export default ({ action }, { services  }) => {
    const { ItemsService } = services;

    action('items.create',  async (meta, { accountability, schema, database }) => {
        const itemsService = new ItemsService('<collection-name>', { accountability, schema, knex: database });

        // itemsService.createOne(...);
    });
};
```

**Reply by TinyTurtleMedia** [reply]
OK, awesome, I think that's enough info to get me started. Thanks!

**Comment by kadlinobit** [comment]
Hi, I'm new to Directus and developing my first API extension (be it a hook or endpoint) was pain in the neck. 

Eventually I have discovered it's much better to omit the extensions SDK at the moment (before there is an "extension dev environment") and just develop your extensions directly in the `/directus-project-folder/extensions/...`

All you have to do is to use commonJS module syntax rather than ES6 module - which is basically just using `module.exports = function()` instead of `export default {}`. 

Extension is then much easier to debug and when there is some kind of extension development environment in the future, you can just rewrite all the extensions to ES6 modules and use the extension SDK to build them properly.

**Reply by nickrum** [reply]
Could you elaborate on which part exactly were painfull? This would help us to improve the workflow until there is an "extension dev environment".

**Reply by kadlinobit** [reply]
Sure. The biggest pain was debugging - since the code gets minified when built in SDK, it was really hard to debug. 
I also needed to make lot of experiments in the debug console, since I don't know all the Directus backend APIs like services etc. 

I think it would help a lot, if: 
- there was a possibility  to disable code minification (maybe there already is?) 
- there was more documentation on all the services and stuff (of course - I know that this all takes time and I'm not blaming anyone for not having these docs yet :))

**Reply by nickrum** [reply]
Thanks, this helps a lot!

> there was a possibility  to disable code minification (maybe there already is?)

There is no option to do that right now. I'd be down to add one, though.

> there was more documentation on all the services and stuff (of course - I know that this all takes time and I'm not blaming anyone for not having these docs yet :))

I fully agree with this. It's already on our radar (ref #5088).

**Reply by TinyTurtleMedia** [reply]
@kadlinobit So I have the below code located in extensions/hooks/functions/index.js. Would you expect that to work? I'm using Cloudron if that's relevant as well. I've tried npx directus start, but it says I need to install npm, which I don't know if I want to do on Cloudron. I've restarted Directus via Cloudron's admin panel, should that be enough to have it detect a working plugin?
```
module.exports = function ({ filter, action }) => {
	filter('items.create', () => {
		console.log('Creating Item!');
	});

	action('items.create', () => {
		console.log('Item created!');
	});
};
```

**Reply by kadlinobit** [reply]
@DrewHasQuestions You have syntax error there on the first line - mixing classic and arrow function definition. It should be:
```js
module.exports = function ({ filter, action }) {
   /// your code ...
}
```
If I make this fix, the extension gets loaded. I don't have any experience with Cloudron so I can't be any help here.

**Reply by TinyTurtleMedia** [reply]
Ah, still can't get the console to log. Can't tell if Directus is seeing the plugin or not, don't know if Cloudron restarting is effective. I might need to circle back once the documentation gets a little more developed, as I don't think I've quite got the skills to figure out these nuances.

**Reply by nickrum** [reply]
@DrewHasQuestions Sorry to hear you are still having troubles. We're here to help! Directus will print a "Loaded extensions: <your-extension-name>" to the console if it could succesfully load your extension. Did you see something like that?

I don't have exterience with Cloudron either, but it's usually easier toto test extensions locally before deploying them to your production environment.

**Reply by TinyTurtleMedia** [reply]
The "Loaded Extensions:" will print to the DOM console, or the commandline console? I don't see it in either place, even with the boilerplate hook extension example. And what's the actual command to restart directus once an extension is added? npx restart directus doesn't seem to work for me. Though I assume a server restart would work too?

**Reply by nickrum** [reply]
The message will be printed to the commandline console.
There is no specific command to restart Directus, you just exit the process and start it again. Though, in version `9.4.3` we added a `EXTENSIONS_AUTO_RELOAD` that will, when set to `true`, instruct Directus to automatically reload any extension that has changed.
What's the structure of your extensions folder? Make sure you put the extension's `index.js` into `<extensions-folder>/hooks/<extension-name>/index.js`.
